# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/DS3001/assignment2`

In [1]:
! git clone https://www.github.com/kaitlinblakeslee/assignment2

Cloning into 'assignment2'...
remote: Enumerating objects: 39, done.
remote: Counting objects: 100% (39/39), done.
remote: Compressing objects: 100% (33/33), done.
remote: Total 39 (delta 10), reused 31 (delta 6), pack-reused 0
Receiving objects: 100% (39/39), 5.48 MiB | 20.39 MiB/s, done.
Resolving deltas: 100% (10/10), done.


In [18]:
! git clone https://www.github.com/kaitlinblakeslee/assignment2
import pandas as pd
import seaborn as sns
import numpy as np
df = pd.read_csv('./assignment2/data/airbnb_hw.csv')
print(df.head())
print(df.dtypes)
vars = ['Beds', 'Property Types']
df['Beds'].describe()

df['Beds_nan'] = df['Beds'].isnull()

pd.crosstab(df['Beds_nan'],df['Property Type'])

gdf = df.loc[df['Beds_nan'] == 0, :]

print(gdf.shape)
print(gdf['Beds'].describe())

gdf = gdf.loc[gdf['Property Type'].isnull() == 0, :]


# gdf is clean

y = gdf['Price']
x = gdf.loc[:, ['Beds', 'Property Type']]





fatal: destination path 'assignment2' already exists and is not an empty directory.
    Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private ro

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1.1: This paper is about 'tidy' data that are easy to clean and manipulate because of their organized structure with each variable being a column, each row being a variable, and each observational unit as a table. These data sets are easier to understand as well.

1.2: The 'tidy data standard' has been designed to make initial data cleaning simpler by instructing early analysis and exploration and by simplifying the data analysis tools that work well together so the coder can focus on the 'interesting' domain problem quicker.

1.3: The sentence comparing datasets to families is implying that, unlike tidy datasets, there is not a standardized way to be a 'messy' dataset just as there are many different ways a family can be messy or unhappy. The sentence about defining variables and observations means that in general there is a lot of overlap between what can be a variable and what can be an observation. In the paper, they use the example that height and weight are normally thought of as variables but change weight to width and now it may seem as if height and width are dimensions from the same variable. This illustrates the point that there can be ambiguity within a dataset.

1.4: Wickham states that values can either be quantitative (numbers) or qualitative (strings) that belong to a variable and an observation. Variables are values that measure an attribute across a set of units. In contrast, observations contain all values that measure the same unit across attributes.

1.5: Section 2.3 outlines a set of standards for 'tidy data' where each variable must form a column, each observation forms a row, and each type of observational unit forms a table.

1.6: The 5 most common problems with messy datasets are columns listed as values not variables, more than one variable stored in one column, variables stored in both rows and columns, different types of observational units listed in the same table, and a single observational unit stored in multiple tables. The data in table 4 is considered 'messy' because variables form both the rows and the columns, and the column headers are values and not variable names. Melting a dataset is converting columns into rows to make a molten dataset so that each column represents a variable and each row signifies an observation (tidy).

1.7: Table 11 is messy because variables are stored in both rows and columns. Additionally, it also has missing values. Table 12b, on the other hand, is tidy because each column represents its own variable, and each row is a separate observation (the data has been stacked or melted).

1.8: The 'chicken-and-egg' problem with tidy data is that the data is only as useful as the tidy tools that work with it that are intrinsically linked to the data itself. In this way, it can be less efficient to independently alter the data structure or tools. In the future, Wickham hopes we will build upon this framework of tidy data to make it even easier to achieve so data can be analyzed more efficiently. Specifically, Wickham wants to design better tools based on human factors and the cognitive side of analysis. Perhaps the future of data wrangling is using these tidy tools to override or ignore the original data representation and choose its own tidy format to optimize performance.


**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [ ]:
#2.1
import numpy as np
import pandas as pd
adf = pd.read_csv('./assignment2/data/airbnb_hw.csv', low_memory = False) #read in airbnb file as csv
print(adf.head()) #view general overview of shape of data and variables (rows and columns)
var1 = 'Price' #rename variable of interest (the price in this case) so it can be cleaned
#print(adf[var1].unique(),'\n') #I wanted to look at what the data looked like for Price
#print(adf[var1].value_counts(), '\n') #I wanted to see the frequency of each price/data point
print(adf.dtypes) #print the data types for each variable - price is 'object'
adf[var1] = adf[var1].str.replace(',', '') #price is string that I want to change into a numeric variable so first I must get rid of separator ','
print(adf[var1].unique(),'\n') #confirm that ',' is gone
print('Before coercion: \n', adf[var1].describe(),'\n') #data type is still object here
adf[var1] = pd.to_numeric(adf[var1], errors='coerce') #coerce data from object to float (numeric) data type
print('After coercion: \n', adf[var1].describe(),'\n') #confirm data type is float
adf[var1+'_nan'] = adf[var1].isnull() # dummy variable for any missing data points (nan)
print('Number of Missings: \n', sum(adf[var1+'_nan']),'\n') #show the number of Missing Data points (there are none)
adf.to_csv('new_airbnbclean.csv') #save cleaned data as new csv



    Host Id Host Since                                Name Neighbourhood   \
0   5162530        NaN     1 Bedroom in Prime Williamsburg       Brooklyn   
1  33134899        NaN     Sunny, Private room in Bushwick       Brooklyn   
2  39608626        NaN                Sunny Room in Harlem      Manhattan   
3       500  6/26/2008  Gorgeous 1 BR with Private Balcony      Manhattan   
4       500  6/26/2008            Trendy Times Square Loft      Manhattan   

  Property Type  Review Scores Rating (bin)        Room Type  Zipcode  Beds  \
0     Apartment                         NaN  Entire home/apt  11249.0   1.0   
1     Apartment                         NaN     Private room  11206.0   1.0   
2     Apartment                         NaN     Private room  10032.0   1.0   
3     Apartment                         NaN  Entire home/apt  10024.0   3.0   
4     Apartment                        95.0     Private room  10036.0   3.0   

   Number of Records  Number Of Reviews Price  Review Scores R

In [ ]:
#2.2
import numpy as np
import pandas as pd
sdf = pd.read_csv('./assignment2/data/sharks.csv', low_memory = False)
print(sdf.head()) #general overview of data
print(sdf.dtypes) #show all the different types of data - Type is an object
var2 = 'Type' #rename variable of interest
#print(sdf[var2].unique()) #look at all of the unique values for Type variable
sdf[var2] = sdf[var2].replace(np.nan, ' ') #replace nan with missing data
sdf[var2] = sdf[var2].replace(('Questionable', 'Unconfirmed', 'Unverified', 'Invalid', 'Under investigation', 'Boatomg', 'Sea Disaster', 'Boat', 'Watercraft', 'Boating', ' '), 'Unknown') #replace anything that is not 'Provoked' or 'Unprovoked' with 'Unknown' because I want type to simply determine whether or not the attack was provoked or not provoked (if this can be determined)
print(sdf[var2].unique()) #confirm that there are only three possible outcomes for Type
sdf.to_csv('cleanshark.csv') #save cleaned data as new csv

   index   Case Number                  Date    Year        Type      Country  \
0      0    2020.02.05           05-Feb-2020  2020.0  Unprovoked          USA   
1      1  2020.01.30.R  Reported 30-Jan-2020  2020.0    Provoked      BAHAMAS   
2      2    2020.01.17           17-Jan-2020  2020.0  Unprovoked    AUSTRALIA   
3      3    2020.01.16           16-Jan-2020  2020.0  Unprovoked  NEW ZEALAND   
4      4    2020.01.13           13-Jan-2020  2020.0  Unprovoked          USA   

              Area               Location                  Activity  \
0             Maui                    NaN  Stand-Up Paddle boarding   
1           Exumas                    NaN                  Floating   
2  New South Wales          Windang Beach                   Surfing   
3        Southland            Oreti Beach                   Surfing   
4   North Carolina  Rodanthe, Dare County                   Surfing   

              Name  ... Unnamed: 246 Unnamed: 247 Unnamed: 248 Unnamed: 249  \
0      

In [ ]:
#2.3 Dummy variable: For the pretrial data covered in the lecture, clean the WhetherDefendantWasReleasedPretrial variable as well as you can, and, in particular, replace missing values with np.nan.
import numpy as np
import pandas as pd
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
tdf = pd.read_csv(url,  low_memory = False) #read in url data #My computer was struggling to read this in, so honestly I wanted to play around more with these variables but kept getting frustrated when it took forever to load.
tdf['Released'] = tdf['WhetherDefendantWasReleasedPretrial'] #shorten name of variable
var3 = 'Released'
#print(tdf[var3].unique()) #examine possible values - 9 represents unknown or missing data
print(tdf[var3].value_counts()) #how much unknown data is there?
tdf[var3] = tdf[var3].replace(9, np.nan) #replace unknown values with nan
tdf[var3+'_nan'] = tdf[var3].isnull()
print('Number of Missings: \n', sum(tdf[var3+'_nan']),'\n') #number of nan values
print(tdf[var3].unique()) #new values
print(tdf[var3].dtypes) #confirm that all values are floats (including the nan)

#2.4
var4 = 'ImposedSentenceAllChargeInContactEvent'
#print(tdf[var4].unique())
var5 = 'SentenceTypeAllChargesAtConvictionInContactEvent'
print(tdf[var5].unique())
print(tdf[var5].value_counts()) #code 4 + code 9 = 9053 = empty ' ' which is unknown
#print(tdf[var4].value_counts())
#tdf[var4] = tdf[var4].replace(' ', np.nan)
print(tdf[var4].value_counts())
tdf['imposedcharge_NA'] = tdf[var4].isnull() #create dummy variable for if there is missing data
#print(sum(tdf['imposedcharge_NA']),'\n') #count number of missing data
print(pd.crosstab(tdf[var4],tdf[var5]),'\n') #cross tab between sentence type and imposed sentence charge
tdf.loc[tdf[var5] == 9, var4] = np.nan # replace with nan if NA
tdf.loc[tdf[var5] == 4, var4] = np.nan # replace with nan if other
tdf.to_csv('new_pretrial_data.csv') #save cleaned data as new csv

1    19154
0     3801
9       31
Name: Released, dtype: int64
Number of Missings: 
 31 

[nan  0.  1.]
float64
[9 0 1 4 2]
4    8779
0    8720
1    4299
2     914
9     274
Name: SentenceTypeAllChargesAtConvictionInContactEvent, dtype: int64
                    9053
0                   4953
12                  1404
.985626283367556    1051
6                    809
                    ... 
49.9712525667351       1
57.0349075975359       1
79.9260780287474       1
42.1642710472279       1
1.6570841889117        1
Name: ImposedSentenceAllChargeInContactEvent, Length: 484, dtype: int64
0 

SentenceTypeAllChargesAtConvictionInContactEvent  0   1  2     4    9
ImposedSentenceAllChargeInContactEvent                               
                                                  0   0  0  8779  274
.0328542094455852                                 0   6  0     0    0
.0657084188911704                                 3  21  0     0    0
.0985626283367557                                 2   6  

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?

5.1. The Census Bureau states that they gather data on race based on 'self-identification' where citizens are able to choose more than one race group. The US Office of Management and Budget (OMB) has 6 categories to choose from: White, Black or African American, American Indian or Alaska Native, Asian, Native Hawaiian or Other Pacific Islander, and Other.

5.2. We gather race data for many reasons. This data provides information about disparities in healthcare, pay, job availability, housing, and more across races. The government can use it to determine where to allocate funds for specific groups that make up large portions of the population. Insurance companies use this data when offering healthcare plans, home insurance, car insurance, etc. In politics, race data draws attention to inequity. It also reminds us in society how diverse our country is and how minority populations are disproportionately worse off in society. Quality is especially important because these issues have large implications, and if the data is incorrect, then incorrect headlines and information will be spread, and unfair funding will be allocated. However, because race does not have a standard defintion and is self-identifiable, it can be hard to measure. This poses the question of how reliable this data actually is.

5.3. The census is an important source of information that statisticians use to aid in policy-making and reform. It is pretty reliable data since each item is looked at individually. However, in this way it is extremely time-consuming, labor intensive, and expensive. In order to increase efficiency, the bureau may reuse data from other agencies. Yet this decreases its reliability, so there is a trade-off. A large critique is that the US census has undercounted groups that are historically undercounted (Blacks, Latinos, Other), while overcounting white and Asian populations (according to PES estimates). This could be because of the ambiguity in the choices where many people who are Hispanic/Latino indicate 'some other race' because they don't identify with the race categories listed on the form. This can negatively affect their funding so adding a specific option for 'Latino' on the form is a potential improvement for the census to gather more reliable data. This extends to other minority groups as well who are included in 'Other'.

5.4. Sex and gender are becoming increasingly more contentious terms. The US Census states that they are distinct. Sex is a biological construct and gender is a social construct. The survey asks how respondents currently identify their sex, and only lists 'male' or 'female' as options. This question wording and the ambiguity of the two concepts makes this data somewhat unreliable as well when trying to determine the sex composition of the entire population. Additionally,there are no questions about gender on the census. Perhaps, asking more specific questions about what gender one was assigned at birth and what they currently identify as would give a better representation of the population.

5.5. There are privacy issues when it comes to cleaning data variables such as sex, gender, sexual identity, race, etc. If there is missing data, it would be easy to assume one's race or gender based on patterns from other survey questions answered. This gives the potential for data to be skewed based on how the data analyst chooses to handle the missing data. Additionally, certain groups of people may be less likely to want to include their gender, transgender for instance, and then these populations will not be included in the data and potentially represented in funding. To combat these issues in privacy and bias, it is essential to be transparent in how one is cleaning the data. Also using multiple different techniques to present the data in different ways is another option. An extremely bad practice would be a lack in security of this sensitive data (it should be kept anonymous). Additionally, using potentially discriminatory trends with data in race, for example, to predict other trends is dangerous and should be avoided.

5.6. If an algorithm was used to predict values for race, gender, and sexual orientation, I would be concerned at how accurate/reliable the data actually is. I also feel like there is a high probability that an algorithm could reinforce stereotypes for gender based on past data. With people being able to change how they identify their gender (or their race too), this data is somewhat unpredictable, and I would not trust an algorithm to collect it. Additionally, if people specifically chose not to include this data, it feels like a violation of privacy to impute and predict this data that they did not want to share in the first place. Overall, this raises a lot of ethical concerns.

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.